pandas/_libs/tslibs/offsets.pyx

```diff
@@ -844,6 +841,42 @@ cpdef int get_day_of_month(datetime other, day_opt) except? -1:
         raise ValueError(day_opt)
 
 
+def _roll_monthday(n, other, compare):
```

Comment:
this can be de-privatized

Reply:
can you cpdef this? and type things, add a doc-string

Reply:
if you have 2 possibilities for types (they must be both either integers or dateimes), then have 2 functions and name it that way, much more readable

Reply:
Sure.  One consideration that might push in the other direction if we want to get rid of duplicate code: `roll_yearday` and `roll_monthday` are basically special cases of `roll_qtrday` with `modby` of `12` and `1`, respectively.  Merging them would mean doing some unnecessary `mod` calls, no idea what the perf hit would be.